In [1]:
import pandas as pd

# 8.1 Parsing Unix timestamps

It's not obvious how to deal with Unix timestamps in pandas -- it took me quite a while to figure this out. The file we're using here is a popularity-contest file I found on my system at `/var/log/popularity-contest`.

Here's an [explanation of how this file works](http://popcon.ubuntu.com/README).

I'm going to hope that nothing in it is sensitive :)

In [2]:
# Read it, and remove the last row
popcon = pd.read_csv('../data/popularity-contest', sep=' ', )[:-1]
popcon

POPULARITY-CONTEST-0 TIME:1387295813  \
0              1387295797      1367633260   
1              1387295796      1354370480   
2              1387295743      1354341275   
3              1387295743      1387224204   
4              1387295742      1354341253   
5              1387295742      1354341275   
6              1387295740      1387224202   
7              1387295740      1354341253   
8              1387295740      1354341253   
9              1387295740      1354341253   
10             1387295740      1354341253   
11             1387295740      1368154184   
12             1387295738      1354341253   
13             1387295737      1383017433   
14             1387295737      1354341253   
15             1387295737      1354341257   
16             1387295737      1372944527   
17             1387295733      1385396738   
18             1387295732      1354341237   
19             1387295729      1354341277   
20             1387295585      1354341256   
21             1387295585      1385395220   
22             1387295585      1354370476   
23             1387295585      1354341255   
24             1387295574      1360434943   
25             1387295574      1370996606   
26             1387295573      1354341256   
27             1387295573      1354341277   
28             1387295569      1354341253   
29             1387295569      1386692557   
...                   ...             ...   
2867                    0               0   
2868                    0               0   
2869                    0               0   
2870                    0               0   
2871                    0               0   
2872                    0               0   
2873                    0               0   
2874                    0               0   
2875                    0               0   
2876                    0               0   
2877                    0               0   
2878                    0               0   
2879                    0               0   
2880                    0               0   
2881                    0               0   
2882                    0               0   
2883                    0               0   
2884                    0               0   
2885                    0               0   
2886                    0               0   
2887                    0               0   
2888                    0               0   
2889                    0               0   
2890                    0               0   
2891                    0               0   
2892                    0               0   
2893                    0               0   
2894                    0               0   
2895                    0               0   
2896                    0               0   

         ID:d9bdc557ae8941c19e95da1c5da786bd  \
0                                  perl-base   
1                                      login   
2                                 libtalloc2   
3                               libwbclient0   
4                                libselinux1   
5                                 libstdc++6   
6                             libpam-winbind   
7                             libpam-modules   
8                        libpam-ck-connector   
9                                 libpam-cap   
10                                  libpam0g   
11                      libpam-gnome-keyring   
12                                   libgcc1   
13                                     libc6   
14                                   libcap2   
15                          libck-connector0   
16                               libdbus-1-3   
17                           fingerprint-gui   
18                                      dash   
19                        popularity-contest   
20                                   mlocate   
21                                      dpkg   
22                                 coreutils   
23                                util-linux   
24            

Rename the columns to ['atime', 'ctime', 'package-name', 'mru-program', 'tag']. The colums are the access time, created time, package name, recently used program, and a tag

In [3]:
popcon = popcon.rename(index=str, columns={'POPULARITY-CONTEST-0':'atime', 'TIME:1387295813':'ctime', 'ID:d9bdc557ae8941c19e95da1c5da786bd':'package-name', 'ARCH:amd64':'mru-program', 'POPCONVER:1.53ubuntu1':'tag'})
print(popcon.dtypes)
popcon

atime           object
ctime           object
package-name    object
mru-program     object
tag             object
dtype: object


atime       ctime                            package-name  \
0     1387295797  1367633260                               perl-base   
1     1387295796  1354370480                                   login   
2     1387295743  1354341275                              libtalloc2   
3     1387295743  1387224204                            libwbclient0   
4     1387295742  1354341253                             libselinux1   
5     1387295742  1354341275                              libstdc++6   
6     1387295740  1387224202                          libpam-winbind   
7     1387295740  1354341253                          libpam-modules   
8     1387295740  1354341253                     libpam-ck-connector   
9     1387295740  1354341253                              libpam-cap   
10    1387295740  1354341253                                libpam0g   
11    1387295740  1368154184                    libpam-gnome-keyring   
12    1387295738  1354341253                                 libgcc1   
13    1387295737  1383017433                                   libc6   
14    1387295737  1354341253                                 libcap2   
15    1387295737  1354341257                        libck-connector0   
16    1387295737  1372944527                             libdbus-1-3   
17    1387295733  1385396738                         fingerprint-gui   
18    1387295732  1354341237                                    dash   
19    1387295729  1354341277                      popularity-contest   
20    1387295585  1354341256                                 mlocate   
21    1387295585  1385395220                                    dpkg   
22    1387295585  1354370476                               coreutils   
23    1387295585  1354341255                              util-linux   
24    1387295574  1360434943                                 upstart   
25    1387295574  1370996606                                    cups   
26    1387295573  1354341256                               findutils   
27    1387295573  1354341277                               logrotate   
28    1387295569  1354341253                               diffutils   
29    1387295569  1386692557                                     apt   
...          ...         ...                                     ...   
2867           0           0                            gnumeric-doc   
2868           0           0                         libplank-common   
2869           0           0                      libopencore-amrwb0   
2870           0           0                      libgweather-common   
2871           0           0                                 libzmq1   
2872           0           0                       libdjvulibre-text   
2873           0           0                           libjtidy-java   
2874           0           0                        libopenlibm-base   
2875           0           0                              libarpack2   
2876           0           0                             libssl0.9.8   
2877           0           0              python-scikits.statsmodels   
2878           0           0                          python-sklearn   
2879           0           0                             libmodplug1   
2880           0           0                             libvlccore5   
2881           0           0                              libgck-1-0   
2882           0           0                           libxcb-icccm4   
2883           0           0                             gsfonts-x11   
2884           0           0                   elementary-wallpapers   
2885           0           0          libboost-program-options1.49.0   
2886           0           0                   gnome-icon-theme-full   
2887           0           0                           libgtkhotkey1   
2888           0           0                           libvncserver0   
2889           0           0                             libspeechd2   
2890           0           0    elementaryos-report

The magical part about parsing timestamps in pandas is that numpy datetimes are already stored as Unix timestamps. So all we need to do is tell pandas that these integers are actually datetimes -- it doesn't need to do any conversion at all.

Start to convert `atime` and `ctime` column as int. Every numpy array and pandas series has a dtype -- this is usually `int64`, `float64`, or `object`.

In [4]:
popcon[['atime', 'ctime']] = popcon[['atime', 'ctime']].astype("int64")
print(popcon.dtypes)
popcon

atime            int64
ctime            int64
package-name    object
mru-program     object
tag             object
dtype: object


atime       ctime                            package-name  \
0     1387295797  1367633260                               perl-base   
1     1387295796  1354370480                                   login   
2     1387295743  1354341275                              libtalloc2   
3     1387295743  1387224204                            libwbclient0   
4     1387295742  1354341253                             libselinux1   
5     1387295742  1354341275                              libstdc++6   
6     1387295740  1387224202                          libpam-winbind   
7     1387295740  1354341253                          libpam-modules   
8     1387295740  1354341253                     libpam-ck-connector   
9     1387295740  1354341253                              libpam-cap   
10    1387295740  1354341253                                libpam0g   
11    1387295740  1368154184                    libpam-gnome-keyring   
12    1387295738  1354341253                                 libgcc1   
13    1387295737  1383017433                                   libc6   
14    1387295737  1354341253                                 libcap2   
15    1387295737  1354341257                        libck-connector0   
16    1387295737  1372944527                             libdbus-1-3   
17    1387295733  1385396738                         fingerprint-gui   
18    1387295732  1354341237                                    dash   
19    1387295729  1354341277                      popularity-contest   
20    1387295585  1354341256                                 mlocate   
21    1387295585  1385395220                                    dpkg   
22    1387295585  1354370476                               coreutils   
23    1387295585  1354341255                              util-linux   
24    1387295574  1360434943                                 upstart   
25    1387295574  1370996606                                    cups   
26    1387295573  1354341256                               findutils   
27    1387295573  1354341277                               logrotate   
28    1387295569  1354341253                               diffutils   
29    1387295569  1386692557                                     apt   
...          ...         ...                                     ...   
2867           0           0                            gnumeric-doc   
2868           0           0                         libplank-common   
2869           0           0                      libopencore-amrwb0   
2870           0           0                      libgweather-common   
2871           0           0                                 libzmq1   
2872           0           0                       libdjvulibre-text   
2873           0           0                           libjtidy-java   
2874           0           0                        libopenlibm-base   
2875           0           0                              libarpack2   
2876           0           0                             libssl0.9.8   
2877           0           0              python-scikits.statsmodels   
2878           0           0                          python-sklearn   
2879           0           0                             libmodplug1   
2880           0           0                             libvlccore5   
2881           0           0                              libgck-1-0   
2882           0           0                           libxcb-icccm4   
2883           0           0                             gsfonts-x11   
2884           0           0                   elementary-wallpapers   
2885           0           0          libboost-program-options1.49.0   
2886           0           0                   gnome-icon-theme-full   
2887           0           0                           libgtkhotkey1   
2888           0           0                           libvncserver0   
2889           0           0                             libspeechd2   
2890           0           0    elementaryos-report

 Some of the time types available are `datetime64[s]`, `datetime64[ms]`, and `datetime64[us]`. There are also `timedelta` types, similarly.

We can use the `pd.to_datetime` function to convert  our integer timestamps into datetimes. This is a constant-time operation -- we're not actually changing any of the data, just how pandas thinks about it.

do it for `atime` and `ctime`

In [5]:
popcon['atime'] = pd.to_datetime(popcon['atime'], unit = 's')
popcon['ctime'] = pd.to_datetime(popcon['ctime'], unit = 's')
print(popcon.dtypes)
popcon

atime           datetime64[ns]
ctime           datetime64[ns]
package-name            object
mru-program             object
tag                     object
dtype: object


atime               ctime  \
0    2013-12-17 15:56:37 2013-05-04 02:07:40   
1    2013-12-17 15:56:36 2012-12-01 14:01:20   
2    2013-12-17 15:55:43 2012-12-01 05:54:35   
3    2013-12-17 15:55:43 2013-12-16 20:03:24   
4    2013-12-17 15:55:42 2012-12-01 05:54:13   
5    2013-12-17 15:55:42 2012-12-01 05:54:35   
6    2013-12-17 15:55:40 2013-12-16 20:03:22   
7    2013-12-17 15:55:40 2012-12-01 05:54:13   
8    2013-12-17 15:55:40 2012-12-01 05:54:13   
9    2013-12-17 15:55:40 2012-12-01 05:54:13   
10   2013-12-17 15:55:40 2012-12-01 05:54:13   
11   2013-12-17 15:55:40 2013-05-10 02:49:44   
12   2013-12-17 15:55:38 2012-12-01 05:54:13   
13   2013-12-17 15:55:37 2013-10-29 03:30:33   
14   2013-12-17 15:55:37 2012-12-01 05:54:13   
15   2013-12-17 15:55:37 2012-12-01 05:54:17   
16   2013-12-17 15:55:37 2013-07-04 13:28:47   
17   2013-12-17 15:55:33 2013-11-25 16:25:38   
18   2013-12-17 15:55:32 2012-12-01 05:53:57   
19   2013-12-17 15:55:29 2012-12-01 05:54:37   
20   2013-12-17 15:53:05 2012-12-01 05:54:16   
21   2013-12-17 15:53:05 2013-11-25 16:00:20   
22   2013-12-17 15:53:05 2012-12-01 14:01:16   
23   2013-12-17 15:53:05 2012-12-01 05:54:15   
24   2013-12-17 15:52:54 2013-02-09 18:35:43   
25   2013-12-17 15:52:54 2013-06-12 00:23:26   
26   2013-12-17 15:52:53 2012-12-01 05:54:16   
27   2013-12-17 15:52:53 2012-12-01 05:54:37   
28   2013-12-17 15:52:49 2012-12-01 05:54:13   
29   2013-12-17 15:52:49 2013-12-10 16:22:37   
...                  ...                 ...   
2867 1970-01-01 00:00:00 1970-01-01 00:00:00   
2868 1970-01-01 00:00:00 1970-01-01 00:00:00   
2869 1970-01-01 00:00:00 1970-01-01 00:00:00   
2870 1970-01-01 00:00:00 1970-01-01 00:00:00   
2871 1970-01-01 00:00:00 1970-01-01 00:00:00   
2872 1970-01-01 00:00:00 1970-01-01 00:00:00   
2873 1970-01-01 00:00:00 1970-01-01 00:00:00   
2874 1970-01-01 00:00:00 1970-01-01 00:00:00   
2875 1970-01-01 00:00:00 1970-01-01 00:00:00   
2876 1970-01-01 00:00:00 1970-01-01 00:00:00   
2877 1970-01-01 00:00:00 1970-01-01 00:00:00   
2878 1970-01-01 00:00:00 1970-01-01 00:00:00   
2879 1970-01-01 00:00:00 1970-01-01 00:00:00   
2880 1970-01-01 00:00:00 1970-01-01 00:00:00   
2881 1970-01-01 00:00:00 1970-01-01 00:00:00   
2882 1970-01-01 00:00:00 1970-01-01 00:00:00   
2883 1970-01-01 00:00:00 1970-01-01 00:00:00   
2884 1970-01-01 00:00:00 1970-01-01 00:00:00   
2885 1970-01-01 00:00:00 1970-01-01 00:00:00   
2886 1970-01-01 00:00:00 1970-01-01 00:00:00   
2887 1970-01-01 00:00:00 1970-01-01 00:00:00   
2888 1970-01-01 00:00:00 1970-01-01 00:00:00   
2889 1970-01-01 00:00:00 1970-01-01 00:00:00   
2890 1970-01-01 00:00:00 1970-01-01 00:00:00   
2891 1970-01-01 00:00:00 1970-01-01 00:00:00   
2892 1970-01-01 00:00:00 1970-01-01 00:00:00   
2893 1970-01-01 00:00:00 1970-01-01 00:00:00   
2894 1970-01-01 00:00:00 1970-01-01 00:00:00   
2895 1970-01-01 00:00:00 1970-01-01 00:00:00   
2896 1970-01-01 00:00:00 1970-01-01 00:00:00   

                                package-name  \
0                                  perl-base   
1                                      login   
2                                 libtalloc2   
3                               libwbclient0   
4                                libselinux1   
5                                 libstdc++6   
6                             libpam-winbind   
7                             libpam-modules   
8                        libpam-ck-connector   
9                                 libpam-cap   
10                                  libpam0g   
11                      libpam-gnome-keyring   
12                                   libgcc1   
13                                     libc6   
14                                   libcap2   
15                          libck-connector0   
16                               libdbus-1-3   
17                           fingerprint-gui   
18                                      dash   
19                        popularity-contest   
20                                

using `dtype` attribute print the type of `atime`. 

In [6]:
popcon['atime'].dtype

dtype('<M8[ns]')

Get rid of everything with timestamp 0. Notice how we can just use a string in this comparison, even though it's actually a timestamp on the inside? That is because pandas is amazing.

In [10]:
popcon = popcon[~((popcon.atime == "1970-01-01 00:00:00") | (popcon.ctime == "1970-01-01 00:00:00"))]
popcon

atime               ctime                 package-name  \
0    2013-12-17 15:56:37 2013-05-04 02:07:40                    perl-base   
1    2013-12-17 15:56:36 2012-12-01 14:01:20                        login   
2    2013-12-17 15:55:43 2012-12-01 05:54:35                   libtalloc2   
3    2013-12-17 15:55:43 2013-12-16 20:03:24                 libwbclient0   
4    2013-12-17 15:55:42 2012-12-01 05:54:13                  libselinux1   
5    2013-12-17 15:55:42 2012-12-01 05:54:35                   libstdc++6   
6    2013-12-17 15:55:40 2013-12-16 20:03:22               libpam-winbind   
7    2013-12-17 15:55:40 2012-12-01 05:54:13               libpam-modules   
8    2013-12-17 15:55:40 2012-12-01 05:54:13          libpam-ck-connector   
9    2013-12-17 15:55:40 2012-12-01 05:54:13                   libpam-cap   
10   2013-12-17 15:55:40 2012-12-01 05:54:13                     libpam0g   
11   2013-12-17 15:55:40 2013-05-10 02:49:44         libpam-gnome-keyring   
12   2013-12-17 15:55:38 2012-12-01 05:54:13                      libgcc1   
13   2013-12-17 15:55:37 2013-10-29 03:30:33                        libc6   
14   2013-12-17 15:55:37 2012-12-01 05:54:13                      libcap2   
15   2013-12-17 15:55:37 2012-12-01 05:54:17             libck-connector0   
16   2013-12-17 15:55:37 2013-07-04 13:28:47                  libdbus-1-3   
17   2013-12-17 15:55:33 2013-11-25 16:25:38              fingerprint-gui   
18   2013-12-17 15:55:32 2012-12-01 05:53:57                         dash   
19   2013-12-17 15:55:29 2012-12-01 05:54:37           popularity-contest   
20   2013-12-17 15:53:05 2012-12-01 05:54:16                      mlocate   
21   2013-12-17 15:53:05 2013-11-25 16:00:20                         dpkg   
22   2013-12-17 15:53:05 2012-12-01 14:01:16                    coreutils   
23   2013-12-17 15:53:05 2012-12-01 05:54:15                   util-linux   
24   2013-12-17 15:52:54 2013-02-09 18:35:43                      upstart   
25   2013-12-17 15:52:54 2013-06-12 00:23:26                         cups   
26   2013-12-17 15:52:53 2012-12-01 05:54:16                    findutils   
27   2013-12-17 15:52:53 2012-12-01 05:54:37                    logrotate   
28   2013-12-17 15:52:49 2012-12-01 05:54:13                    diffutils   
29   2013-12-17 15:52:49 2013-12-10 16:22:37                          apt   
...                  ...                 ...                          ...   
2068 2011-12-28 05:43:33 2012-12-01 05:54:22                  ibus-pinyin   
2069 2011-12-25 03:26:53 2012-12-01 05:54:16                      toshset   
2070 2011-12-18 23:46:08 2012-12-01 05:54:20         printer-driver-c2esp   
2071 2011-12-17 15:48:53 2012-12-01 05:54:37                avahi-autoipd   
2072 2011-12-17 15:48:25 2012-12-01 05:54:35         libavahi-common-data   
2073 2011-12-13 17:11:10 2012-12-01 05:54:37                    pppconfig   
2074 2011-12-12 13:47:15 2012-12-01 05:54:37                  base-passwd   
2075 2011-12-07 10:26:37 2012-12-01 05:54:14                   libgksu2-0   
2076 2011-12-05 16:01:44 2012-12-01 05:54:15                       ltrace   
2077 2011-12-05 11:35:27 2012-12-01 05:54:57            libclass-isa-perl   
2078 2011-11-27 10:00:49 2012-12-01 05:54:27                   memtest86+   
2079 2011-11-08 10:47:49 2012-12-01 05:54:37                    dmidecode   
2080 2011-10-17 12:25:55 2012-12-01 05:54:57               libswitch-perl   
2081 2011-10-17 12:04:15 2012-12-01 05:54:57        libfile-mimeinfo-perl   
2082 2011-10-17 12:03:46 2012-12-01 05:54:57    libfile-desktopentry-perl   
2083 2011-10-17 12:00:54 2012-12-01 05:54:57      libcrypt-passwdmd5-perl   
2084 2011-08-25 17:53:02 2012-12-01 05:54:22               guile-1.8-libs   
2085 2011-08-16 11:10:34 2012-12-01 05:54:37        grub-gfxpayload-lists   
2086 2011-08-12 20:06:55 2012-12-01 05:54:37                         tcpd   
2087 2011-07-28 20:05:14 2012-12-01 05:54:15                     mtr-tiny   
2088 2011-07-03

Filter out package-name which are lib, ie containing lib in their name. You can use the `~` operator for "not" 

In [13]:
nonlibraries = popcon[~popcon["package-name"].str.contains('lib')]
nonlibraries

atime               ctime                 package-name  \
0    2013-12-17 15:56:37 2013-05-04 02:07:40                    perl-base   
1    2013-12-17 15:56:36 2012-12-01 14:01:20                        login   
17   2013-12-17 15:55:33 2013-11-25 16:25:38              fingerprint-gui   
18   2013-12-17 15:55:32 2012-12-01 05:53:57                         dash   
19   2013-12-17 15:55:29 2012-12-01 05:54:37           popularity-contest   
20   2013-12-17 15:53:05 2012-12-01 05:54:16                      mlocate   
21   2013-12-17 15:53:05 2013-11-25 16:00:20                         dpkg   
22   2013-12-17 15:53:05 2012-12-01 14:01:16                    coreutils   
23   2013-12-17 15:53:05 2012-12-01 05:54:15                   util-linux   
24   2013-12-17 15:52:54 2013-02-09 18:35:43                      upstart   
25   2013-12-17 15:52:54 2013-06-12 00:23:26                         cups   
26   2013-12-17 15:52:53 2012-12-01 05:54:16                    findutils   
27   2013-12-17 15:52:53 2012-12-01 05:54:37                    logrotate   
28   2013-12-17 15:52:49 2012-12-01 05:54:13                    diffutils   
29   2013-12-17 15:52:49 2013-12-10 16:22:37                          apt   
30   2013-12-17 15:52:49 2012-12-01 05:54:15                  debianutils   
31   2013-12-17 05:11:46 2012-12-01 05:54:09                         crda   
32   2013-12-17 05:11:46 2012-12-01 05:54:09               wireless-regdb   
54   2013-12-17 04:55:44 2012-12-01 05:54:14                 install-info   
55   2013-12-17 04:55:42 2013-02-09 18:49:39                       man-db   
56   2013-12-17 04:55:42 2012-12-01 05:54:57                      debconf   
57   2013-12-17 04:55:39 2013-12-17 04:55:42                          ddd   
58   2013-12-17 04:55:39 2012-12-01 05:53:59                          tar   
61   2013-12-17 04:55:38 2013-05-04 02:07:37                         perl   
64   2013-12-17 04:55:38 2012-12-01 05:53:57                         gzip   
66   2013-12-17 04:55:38 2013-12-10 16:24:47                    apt-utils   
68   2013-12-17 04:47:50 2012-12-01 05:54:16                   groff-base   
72   2013-12-17 04:44:40 2012-12-01 05:54:58             python-gobject-2   
73   2013-12-17 04:44:40 2012-12-01 15:41:24            virtualenvwrapper   
74   2013-12-17 04:44:40 2013-11-25 16:19:55    python-lazr.restfulclient   
...                  ...                 ...                          ...   
2056 2012-01-18 06:12:07 2012-12-01 05:54:36  xserver-xorg-video-neomagic   
2057 2012-01-18 06:03:27 2012-12-01 05:54:36      xserver-xorg-video-r128   
2058 2012-01-18 05:19:45 2012-12-01 05:54:36       xserver-xorg-video-mga   
2059 2012-01-18 05:17:19 2012-12-01 05:54:36    xserver-xorg-video-mach64   
2060 2012-01-18 05:13:07 2012-12-01 05:54:36    xserver-xorg-video-cirrus   
2061 2012-01-18 03:54:18 2012-12-01 05:54:37     xserver-xorg-input-mouse   
2062 2012-01-08 11:39:26 2012-12-01 05:54:13                       espeak   
2063 2012-01-03 15:42:49 2012-12-01 05:54:13                           iw   
2064 2011-12-31 03:01:58 2012-12-01 05:54:28               python-virtkey   
2065 2011-12-31 02:58:00 2012-12-01 05:54:28       python-reportlab-accel   
2066 2011-12-31 02:57:59 2012-12-01 05:54:28              python-renderpm   
2067 2011-12-29 10:21:09 2012-12-01 05:54:15                         lshw   
2068 2011-12-28 05:43:33 2012-12-01 05:54:22                  ibus-pinyin   
2069 2011-12-25 03:26:53 2012-12-01 05:54:16                      toshset   
2070 2011-12-18 23:46:08 2012-12-01 05:54:20         printer-driver-c2esp   
2071 2011-12-17 15:48:53 2012-12-01 05:54:37                avahi-autoipd   
2073 2011-12-13 17:11:10 2012-12-01 05:54:37                    pppconfig   
2074 2011-12-12 13:47:15 2012-12-01 05:54:37                  base-passwd   
2076 2011-12-05 16:01:44 2012-12-01 05:54:15                       ltrace   
2078 2011-11-27 10:00:49 2012-12-01 05:54:27                   memtest86+   
2079 2011-11-08

sort result by ctime

In [15]:
nonlibraries.sort_values('ctime', ascending=False)[:10]

atime               ctime               package-name  \
57  2013-12-17 04:55:39 2013-12-17 04:55:42                        ddd   
450 2013-12-16 20:03:20 2013-12-16 20:05:13                     nodejs   
454 2013-12-16 20:03:20 2013-12-16 20:05:04  switchboard-plug-keyboard   
445 2013-12-16 20:03:20 2013-12-16 20:05:04      thunderbird-locale-en   
396 2013-12-16 20:08:27 2013-12-16 20:05:03            software-center   
449 2013-12-16 20:03:20 2013-12-16 20:05:00           samba-common-bin   
397 2013-12-16 20:08:25 2013-12-16 20:04:59      postgresql-client-9.1   
398 2013-12-16 20:08:23 2013-12-16 20:04:58             postgresql-9.1   
452 2013-12-16 20:03:20 2013-12-16 20:04:55                   php5-dev   
440 2013-12-16 20:03:20 2013-12-16 20:04:54                   php-pear   

                                           mru-program             tag  
57                                        /usr/bin/ddd  <RECENT-CTIME>  
450                                       /usr/bin/npm  <RECENT-CTIME>  
454       /usr/lib/plugs/pantheon/keyboard/options.txt  <RECENT-CTIME>  
445  /usr/lib/thunderbird-addons/extensions/langpac...  <RECENT-CTIME>  
396                   /usr/sbin/update-software-center  <RECENT-CTIME>  
449                                /usr/bin/net.samba3  <RECENT-CTIME>  
397                   /usr/lib/postgresql/9.1/bin/psql  <RECENT-CTIME>  
398             /usr/lib/postgresql/9.1/bin/postmaster  <RECENT-CTIME>  
452                  /usr/include/php5/main/snprintf.h  <RECENT-CTIME>  
440                        /usr/share/php/XML/Util.php  <RECENT-CTIME>

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  